### PR TITLE
Dedup trust domain aliases for TLS validation context

### PR DESCRIPTION
**Pull request by kailun-qin**
Fixes https://github.com/istio/istio/issues/32245

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Review by lei-tang — APPROVED** [review]

**Comment by lei-tang** [comment]
Thanks for the fix!

File: pilot/pkg/security/model/authentication.go [hunk 1]
```diff
@@ -268,7 +270,8 @@ func ApplyToCommonTLSContext(tlsContext *tls.CommonTlsContext, proxy *model.Prox
 	// TODO: if user explicitly specifies SANs - should we alter his explicit config by adding all spifee aliases?
 	matchSAN := util.StringToExactMatch(subjectAltNames)
 	if len(trustDomainAliases) > 0 {
-		matchSAN = append(matchSAN, util.StringToPrefixMatch(appendURIPrefixToTrustDomain(trustDomainAliases))...)
+		matchSAN = append(matchSAN, util.StringToPrefixMatch(
+			appendURIPrefixToTrustDomain(dedupTrustDomains(trustDomainAliases)))...)
```
Comment (incfly) [inline pilot/pkg/security/model/authentication.go:274]: I'm thinking shall we do it even earlier? Can you look a bit upper of the call chain of `ApplyToCommonTLSContext` see if there's an appropriate place?

That way, other future callers does not need to memorize calling (or implmementing) this dedup function.

**Comment by kailun-qin** [comment]
> Maybe it is more efficient to just do this one time in
> 
> https://github.com/istio/istio/blob/9112214fba47f9f228b8dd83551b231602f3c72b/pilot/pkg/networking/plugin/authn/util.go#L26
> 
> ?

Thanks for the hint!

Yes, I agree this would be more efficient and may also bring about some benefits that @incfly may think of - other calls use `TrustDomainForValidation` need not to dedup once more . The only concern is that if we dedup at a upper layer, we cannot guarantee every call into `ApplyToCommonTLSContext` has `trustDomainAliases` input deduped.

But for now the only applicable case (actually using td for validation) is `InboundMTLSConfiguration`, where a dedup here is enough https://github.com/istio/istio/blob/9112214fba47f9f228b8dd83551b231602f3c72b/pilot/pkg/networking/plugin/authn/util.go#L26.

Updated to dedup when try to get trust domains for validation from the mesh config. Let me know if any comment. Thanks!
@howardjohn @lei-tang @incfly

File: pilot/pkg/security/model/authentication.go [hunk 2]
```diff
@@ -285,6 +288,21 @@ func ApplyToCommonTLSContext(tlsContext *tls.CommonTlsContext, proxy *model.Prox
...
+func dedupTrustDomains(tds []string) []string {
+	known := sets.NewSet()
+	deduped := []string{}
+
+	for _, td := range tds {
+		if !known.Contains(td) {
+			known.Insert(td)
+			deduped = append(deduped, td)
+		} else {
+			log.Debugf("skipping duplicated trust domain %s", td)
```
[the remark this answers is not in the document]
Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go]: Removed, thanks!

File: pilot/pkg/security/model/authentication.go (hunk 1 above)
Comment (kailun-qin) [reply pilot/pkg/security/model/authentication.go]: Thanks for the comment! Moved to where John suggests https://github.com/istio/istio/pull/32265#pullrequestreview-639223896. Please kindly see the details below.

**Comment by kailun-qin** [comment]
Adding hold to merge label... let's see if any further comment from @howardjohn @lei-tang @incfly 
Thanks!

**Review by howardjohn — APPROVED** [review]

**Review by incfly — APPROVED** [review]

**Pull request merged**